cartographer/mapping/internal/constraints/constraint_builder_2d.cc

```diff
@@ -77,106 +80,112 @@ void ConstraintBuilder2D::MaybeAddConstraint(
       options_.max_constraint_distance()) {
     return;
   }
-  if (sampler_.Pulse()) {
-    common::MutexLocker locker(&mutex_);
-    constraints_.emplace_back();
-    kQueueLengthMetric->Set(constraints_.size());
-    auto* const constraint = &constraints_.back();
-    ++pending_computations_[current_computation_];
-    const int current_computation = current_computation_;
-    ScheduleSubmapScanMatcherConstructionAndQueueWorkItem(
-        submap_id, submap->grid(), [=]() EXCLUDES(mutex_) {
-          ComputeConstraint(submap_id, submap, node_id,
-                            false, /* match_full_submap */
-                            constant_data, initial_relative_pose, constraint);
-          FinishComputation(current_computation);
-        });
+  if (!sampler_.Pulse()) return;
+
+  common::MutexLocker locker(&mutex_);
+  if (when_done_) {
+    LOG(WARNING)
+        << "MaybeAddConstraint was called while WhenDone was scheduled.";
   }
+  constraints_.emplace_back();
+  kQueueLengthMetric->Set(constraints_.size());
+  auto* const constraint = &constraints_.back();
+  const auto* scan_matcher =
+      DispatchScanMatcherConstruction(submap_id, submap->grid());
+  auto constraint_task = common::make_unique<common::Task>();
+  constraint_task->SetWorkItem([=]() EXCLUDES(mutex_) {
+    ComputeConstraint(submap_id, submap, node_id, false, /* match_full_submap */
+                      constant_data, initial_relative_pose, *scan_matcher,
+                      constraint);
+  });
```

Comment:
that's a weird formatting, is it with clang-format?

Reply:
Yeah, it's clang-formatted.